src/async_impl/response.rs

```diff
@@ -106,16 +121,53 @@ impl fmt::Debug for Response {
     }
 }
 
+impl decoder::PollChunk for Body {
+    fn poll_chunk(&mut self) -> io::Result<Option<Chunk>> {
+        match self.poll() {
+            Ok(Async::Ready(Some(chunk))) => Ok(Some(chunk)),
+            Ok(Async::Ready(None)) => Ok(None),
+            Ok(Async::NotReady) => Err(io::ErrorKind::WouldBlock.into()),
+            Err(e) => Err(::error::into_io(e)),
+        }
+    }
+}
+
+pub struct DecodedBody {
```

Comment:
So, my opinion may seem odd, and if so, that's OK. If it seems too absurd, let me know.

But, here's an API that I think would make sense to expose:

```rust
impl Response {
    pub fn body_mut(&mut self) -> &mut Decoder {
        // ...
    }
}

struct Decoder {
    inner: DecoderImplWhateverNameNotSuperImportant,
}

impl Stream for Decoder {
    type Item = Chunk;
    type Error = io::Error;
}
```

I realize that libflate needs an `io::Read`, but I'm not certain yet about exposing those traits to the public.

The inner decoder impl can provide a `Read` for libflate to use, which does like you've done, just polling a chunk out when needed. When the libflate decoder returns some decoded bytes, that can be emitted as a new `Chunk`.